### DISCUSSION TITLE
Updated to 5.3.0 and having some issues

### DISCUSSION THREAD

**Discussion by Zer0x00** [celery/celery#8314] [Q&A]
First of all, it's worth mentioning that the same configuration worked without any issues in version 5.2.7. 
I'm using Python 3.11.4 (the version has not been changed) and I updated celery to version 5.3.0.

When starting celery, two issues occur:
- Celery beat throws errors
- I receive a deprecation warning for "broker_connection_retry" even though I'm not using it anywhere

```python
dev@527b01265e12:/example-app$ celery -A app.worker worker --loglevel=INFO --beat
 -------------- celery@527b01265e12 v5.3.0 (emerald-rush)
--- ***** ----- 
-- ******* ---- Linux-5.15.90.1-microsoft-standard-WSL2-x86_64-with-glibc2.35 2023-06-13 10:57:21
- *** --- * --- 
- ** ---------- [config]
- ** ---------- .> app:         example-app:0x7fa7e7ade400
- ** ---------- .> transport:   redis://host.docker.internal:6379/0
- ** ---------- .> results:     postgresql://postgres:**@host.docker.internal/dci
- *** --- * --- .> concurrency: 5 (prefork)
-- ******* ---- .> task events: OFF (enable -E to monitor tasks in this worker)
--- ***** ----- 
 -------------- [queues]
                .> example-app           exchange=example-app(direct) key=example-app

[tasks]
  . task1

[2023-06-13 10:57:24,531: WARNING/MainProcess] /usr/local/lib/python3.11/dist-packages/celery/worker/consumer/consumer.py:498: CPendingDeprecationWarning: The broker_connection_retry configuration setting will no longer determine
whether broker connection retries are made during startup in Celery 6.0 and above.
If you wish to retain the existing behavior for retrying connections on startup,
you should set broker_connection_retry_on_startup to True.
  warnings.warn(

[2023-06-13 10:57:24,617: INFO/MainProcess] Connected to redis://host.docker.internal:6379/0
[2023-06-13 10:57:24,624: WARNING/MainProcess] /usr/local/lib/python3.11/dist-packages/celery/worker/consumer/consumer.py:498: CPendingDeprecationWarning: The broker_connection_retry configuration setting will no longer determine
whether broker connection retries are made during startup in Celery 6.0 and above.
If you wish to retain the existing behavior for retrying connections on startup,
you should set broker_connection_retry_on_startup to True.
  warnings.warn(

[2023-06-13 10:57:24,629: INFO/MainProcess] mingle: searching for neighbors
[2023-06-13 10:57:24,735: INFO/Beat] beat: Starting...
[2023-06-13 10:57:24,775: ERROR/Beat] Process Beat
Traceback (most recent call last):
  File "/usr/lib/python3.11/shelve.py", line 111, in __getitem__
    value = self.cache[key]
            ~~~~~~~~~~^^^^^
KeyError: 'entries'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.11/dist-packages/billiard/process.py", line 323, in _bootstrap
    self.run()
  File "/usr/local/lib/python3.11/dist-packages/celery/beat.py", line 715, in run
    self.service.start(embedded_process=True)
  File "/usr/local/lib/python3.11/dist-packages/celery/beat.py", line 631, in start
    humanize_seconds(self.scheduler.max_interval))
                     ^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/kombu/utils/objects.py", line 32, in __get__
    return super().__get__(instance, owner)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/functools.py", line 1001, in __get__
    val = self.func(instance)
          ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/celery/beat.py", line 674, in scheduler
    return self.get_scheduler()
           ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/celery/beat.py", line 665, in get_scheduler
    return symbol_by_name(self.scheduler_cls, aliases=aliases)(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/celery/beat.py", line 510, in __init__
    super().__init__(*args, **kwargs)
  File "/usr/local/lib/python3.11/dist-packages/celery/beat.py", line 264, in __init__
    self.setup_schedule()
  File "/usr/local/lib/python3.11/dist-packages/celery/beat.py", line 538, in setup_schedule
    self._create_schedule()
  File "/usr/local/lib/python3.11/dist-packages/celery/beat.py", line 567, in _create_schedule
    self._store['entries']
    ~~~~~~~~~~~^^^^^^^^^^^
  File "/usr/lib/python3.11/shelve.py", line 114, in __getitem__
    value = Unpickler(f).load()
            ^^^^^^^^^^^^^^^^^^^
ModuleNotFoundError: No module named 'pytz'
[2023-06-13 10:57:24,787: WARNING/Beat] Process Beat:
[2023-06-13 10:57:24,788: WARNING/Beat] Traceback (most recent call last):
[2023-06-13 10:57:24,788: WARNING/Beat]   File "/usr/lib/python3.11/shelve.py", line 111, in __getitem__
    value = self.cache[key]
            ~~~~~~~~~~^^^^^
[2023-06-13 10:57:24,788: WARNING/Beat] KeyError: 'entries'
[2023-06-13 10:57:24,789: WARNING/Beat] 
During handling of the above exception, another exception occurred:
[2023-06-13 10:57:24,789: WARNING/Beat] Traceback (most recent call last):
[2023-06-13 10:57:24,789: WARNING/Beat]   File "/usr/local/lib/python3.11/dist-packages/billiard/process.py", line 323, in _bootstrap
    self.run()
[2023-06-13 10:57:24,789: WARNING/Beat]   File "/usr/local/lib/python3.11/dist-packages/celery/beat.py", line 715, in run
    self.service.start(embedded_process=True)
[2023-06-13 10:57:24,789: WARNING/Beat]   File "/usr/local/lib/python3.11/dist-packages/celery/beat.py", line 631, in start
    humanize_seconds(self.scheduler.max_interval))
                     ^^^^^^^^^^^^^^
[2023-06-13 10:57:24,790: WARNING/Beat]   File "/usr/local/lib/python3.11/dist-packages/kombu/utils/objects.py", line 32, in __get__
    return super().__get__(instance, owner)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2023-06-13 10:57:24,790: WARNING/Beat]   File "/usr/lib/python3.11/functools.py", line 1001, in __get__
    val = self.func(instance)
          ^^^^^^^^^^^^^^^^^^^
[2023-06-13 10:57:24,790: WARNING/Beat]   File "/usr/local/lib/python3.11/dist-packages/celery/beat.py", line 674, in scheduler
    return self.get_scheduler()
           ^^^^^^^^^^^^^^^^^^^^
[2023-06-13 10:57:24,790: WARNING/Beat]   File "/usr/local/lib/python3.11/dist-packages/celery/beat.py", line 665, in get_scheduler
    return symbol_by_name(self.scheduler_cls, aliases=aliases)(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
[2023-06-13 10:57:24,790: WARNING/Beat]   File "/usr/local/lib/python3.11/dist-packages/celery/beat.py", line 510, in __init__
    super().__init__(*args, **kwargs)
[2023-06-13 10:57:24,790: WARNING/Beat]   File "/usr/local/lib/python3.11/dist-packages/celery/beat.py", line 264, in __init__
    self.setup_schedule()
[2023-06-13 10:57:24,790: WARNING/Beat]   File "/usr/local/lib/python3.11/dist-packages/celery/beat.py", line 538, in setup_schedule
    self._create_schedule()
[2023-06-13 10:57:24,790: WARNING/Beat]   File "/usr/local/lib/python3.11/dist-packages/celery/beat.py", line 567, in _create_schedule
    self._store['entries']
    ~~~~~~~~~~~^^^^^^^^^^^
[2023-06-13 10:57:24,790: WARNING/Beat]   File "/usr/lib/python3.11/shelve.py", line 114, in __getitem__
    value = Unpickler(f).load()
            ^^^^^^^^^^^^^^^^^^^
[2023-06-13 10:57:24,790: WARNING/Beat] ModuleNotFoundError: No module named 'pytz'
[2023-06-13 10:57:25,664: INFO/MainProcess] mingle: all alone
[2023-06-13 10:57:25,681: INFO/MainProcess] celery@527b01265e12 ready.
```

**Comment by auvipy** [comment]
ModuleNotFoundError: No module named 'pytz'

did you install it?

**Comment by Zer0x00** [comment]
No I didn't but I saw this pull request which removed the dependency: https://github.com/celery/celery/pull/8159
and since I'm installing it via pip should it not be auto-installed if it is required?

**Reply by auvipy** [reply]
you might need to check your projects code properly.

**Reply by Zer0x00** [reply]
What do you mean? Installing pytz seems to solve the issue but again: should it not be automatically installed by pip?
The deprecation warning remains.

**Reply by gromsterus** [reply]
@Zer0x00 faced the same problem. Solved by deleting the `celerybeat-schedule` file.

**Comment by hofrob** [comment]
To boil this down to a readable report:

>celery/worker/consumer/consumer.py:498: CPendingDeprecationWarning: The broker_connection_retry configuration setting will no longer determine whether broker connection retries are made during startup in Celery 6.0 and above. If you wish to retain the existing behavior for retrying connections on startup, you should set broker_connection_retry_on_startup to True.

This deprecation warning is probably triggered on every installation that does not touch this config variable. Since it's [enabled by default](https://docs.celeryq.dev/en/stable/userguide/configuration.html#std-setting-broker_connection_retry).

# versions

- Celery: 5.3.0
- Python: 3.11

**Reply by ghost** [reply]
Experiencing the same issue. I was able to temporarily solve by:
```python3
celery = Celery(
    __name__,
    broker=redis_addr,
    backend=redis_addr,
    broker_connection_retry=False,
    broker_connection_retry_on_startup=True,
    broker_connection_max_retries=10,
)
```

**Reply by syserr0r** [reply]
How does this work if I want both `broker_connection_retry=True` and `broker_connection_retry_on_startup=True`?

Am I correct that I just need to ignore the deprecation warning?

**Reply by ghost** [reply]
@syserr0r I believe `broker_connection_retry` [is indeed being deprecated (see the red warning)](https://docs.celeryq.dev/en/stable/userguide/configuration.html#broker-connection-retry) and will be removed in future versions (however, for some reason, recent versions are defaulting to `broker_connection_retry=True` even though it is being deprecated, probably to give time for the warning to reach devs and them to modify the code to use the new `broker_connection_retry_on_startup`). You should follow the pattern above for `broker_connection_retry=False, broker_connection_retry_on_startup=True`

**Reply by syserr0r** [reply]
My understanding is they are 2 different settings, and that `broker_connection_retry` has had its functionality split into 2, hence the warning (emphasis mine):

> The broker_connection_retry configuration setting will no longer determine whether broker connection retries are made **_during startup_** in Celery 6.0 and above. If you wish to refrain from retrying connections **_on startup_**, you should set broker_connection_retry_on_startup to False instead.
> 

From the documentation you linked:

`broker_connection_retry` will (emphasis mine):
> Automatically try to re-establish the connection to the AMQP broker **_if lost after the initial connection is made_**.

`broker_connection_retry_on_startup` will (emphasis mine):
> Automatically try to establish the connection to the AMQP broker **_on Celery startup if it is unavailable_**.

What do you think?

**Reply by ghost** [reply]
@syserr0r good point, it looks like `broker_connection_retry` has been forked into two settings:
1. `broker_connection_retry_on_startup` which I mention above, but as you point out, this is only during startup
2. `broker_channel_error_retry` ([link](https://docs.celeryq.dev/en/stable/userguide/configuration.html#broker-channel-error-retry)) which was introduced in 5.3 (same as `broker_connection_retry_on_startup`) and seems to accomplish what you're looking for 

* I would use both `broker_connection_retry_on_startup=True` and `broker_channel_error_retry=True` in your case

```
New in version 5.3.

Default: Disabled.

Automatically try to re-establish the connection to the AMQP broker if any invalid response has been returned.

The retry count and interval is the same as that of broker_connection_retry. Also, this option doesn’t work when broker_connection_retry is False.
```

EDIT: however, given that `this option doesn’t work when broker_connection_retry is False`, I'm not sure how to achieve what you're looking for without `broker_connection_retry=True`

**Reply by syserr0r** [reply]
Given, as you point out, it says for `broker_connection_retry_on_startup`:
> Also, this option doesn’t work when broker_connection_retry is False.

I would assume the following configurations are available:
- `broker_connection_retry` is `False` (`broker_connection_retry_on_startup` is also effectively `False` due to the above)
  - No retries are attempted
- `broker_connection_retry` is `True` and `broker_connection_retry_on_startup` is `False`
  - Retries are attempted only after a successful startup
- `broker_connection_retry` is `True` and `broker_connection_retry_on_startup` is `True`
  - Retries are attempted in all cases, start-up and after

If this is correct it means that while your fix above does prevent the deprecation warning, **_it effectively disables all retries_**!

**Reply by ghost** [reply]
@syserr0r `broker_connection_retry_on_startup` does **not** appear to be disabled when `broker_connection_retry` is `False` (but I could be wrong, the documentation makes no mention):
```
broker_connection_retry_on_startup
Default: Enabled.

Automatically try to establish the connection to the AMQP broker on Celery startup if it is unavailable.

The time between retries is increased for each retry, and is not exhausted before [broker_connection_max_retries](https://docs.celeryq.dev/en/stable/userguide/configuration.html#std-setting-broker_connection_max_retries) is exceeded.
```

Note that given the warning message (`If you wish to refrain from retrying connections on startup, you should set broker_connection_retry_on_startup to False instead`), this implies to me that `broker_connection_retry_on_startup` can be true and `broker_connection_retry` can be false, and celery will still attempt to retry start-up connections. 

<img width="753" alt="Screenshot 2023-07-19 at 4 15 11 PM" src="https://github.com/celery/celery/assets/127700939/a263e326-9c3b-4f5d-996b-97dfd13a0e79">

My Edit referred only to `broker_channel_error_retry`, not to `broker_connection_retry_on_startup`. 

I believe there is an illogical nature to the fact that `broker_channel_error_retry` was introduced in `v5.3`, `broker_connection_retry` was deprecated in `v5.3`, and `broker_channel_error_retry` does not work if `broker_connection_retry` is false. 

TLDR; you can still use `broker_connection_retry_on_startup=True` combined with `broker_connection_retry=False`.

**Reply by syserr0r** [reply]
Good to know the documentation is probably referring to `broker_channel_error_retry`, you are probably correct in that case.

Even still it is worth being aware that setting `broker_connection_retry=False` and `broker_connection_retry_on_startup=True` as you suggest above, may very well only ensure connection retries are made on initial start and not once the broker has started ❗ .
